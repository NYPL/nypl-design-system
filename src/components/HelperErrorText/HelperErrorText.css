.helper-text {
  font-size: var(--font-size--1);
  margin-bottom: var(--space-xs);

  &--error {
    color: var(--ui-error);
  }
}
