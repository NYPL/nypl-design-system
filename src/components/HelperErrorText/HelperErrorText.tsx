import * as React from "react";
import { Box, useStyleConfig } from "@chakra-ui/react";

import generateUUID from "../../helpers/generateUUID";

export type AriaLiveValues = "assertive" | "off" | "polite";

interface HelperErrorTextProps {
  /** Aria attribute. When true, assistive technologies will
   * read the entire DOM element. When false, only changes (additionals or
   * removals) will be read. True by default. */
  ariaAtomic?: boolean;
  /** Aria attribute only used in the invalid state to read error text. This
   * indicates the priority of the text and when it should be presented to users
   * using screen readers; "off" indicates that the content should not be presented,
   * "polite" that it will be announced at the next available time slot, and
   * "assertive" that it should be announced immediately. "polite" by default. */
  ariaLive?: AriaLiveValues;
  /** Additional className to add. */
  className?: string;
  /** Unique ID for accessibility purposes. */
  id?: string;
  /** Toggles between helper and invalid styling. */
  isInvalid?: boolean;
}

/**
 * Helper or Error text for forms
 */
export default function HelperErrorText(
  props: React.PropsWithChildren<HelperErrorTextProps>
) {
  const {
    ariaAtomic = true,
    ariaLive = "polite",
    children,
    className = "",
    id = generateUUID(),
    isInvalid = false,
  } = props;
  // Only announce the text in the invalid state.
  const announceAriaLive = isInvalid;
  const styles = useStyleConfig("HelperErrorText", { isInvalid });

  return (
    <Box
      id={id}
      className={className}
      aria-atomic={ariaAtomic}
      data-isinvalid={isInvalid}
      aria-live={announceAriaLive ? ariaLive : "off"}
      __css={styles}
    >
      {children}
    </Box>
  );
}
