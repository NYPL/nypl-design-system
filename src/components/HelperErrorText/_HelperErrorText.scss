.helper-text {
  @include space-stack-xs;

  font-size: var(--font-size--1);

  &--error {
    color: var(--ui-error);
  }
}
