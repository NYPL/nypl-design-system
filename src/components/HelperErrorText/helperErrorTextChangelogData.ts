/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "2024-09-19",
    version: "3.3.2",
    type: "Bug Fix",
    affects: ["Styles"],
    notes: [
      "Fixes issue where, if text value was not a string, necessary styles weren't applied.",
    ],
  },
  {
    date: "2024-07-25",
    version: "3.2.0",
    type: "Update",
    affects: ["Functionality"],
    notes: ["Exports the `HelperErrorTextProps` interface."],
  },
  {
    date: "2024-03-14",
    version: "3.0.0",
    type: "Update",
    affects: ["Styles"],
    notes: ["Chakra 2.8 update."],
  },
];
