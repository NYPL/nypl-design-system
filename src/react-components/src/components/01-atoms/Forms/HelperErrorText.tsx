/* MT-51 Helper and Error text*/
import * as React from "react";
import bem from "../../../utils/bem";

export interface HelperErrorTextOptions {
  content?: JSX.Element;

  id?: string;
  isError?: boolean;
  modifiers?: string[];
  blockName?: string;
  baseClass?: string;
}

interface HelperErrorTextProps {
  /** unique ID for helper */
  id: string;
  /** Toggles between helper and error styling */
  isError: boolean;
  /** Used for BEM css convention. */
  modifiers?: string[];
  /** Used for BEM css convention. */
  blockName?: string;
  /** Used for BEM css convention. */
  baseClass?: string;
}

export default function HelperErrorText(props: React.PropsWithChildren<HelperErrorTextProps>) {
  const { id, blockName = "input-description", baseClass = "form-item", isError } = props;
  let modifiers = [];

  if (isError) {
    modifiers.push("error");
  }
  if (props.modifiers) {
    modifiers.push(...props.modifiers);
  }
  return <span id={id} className={bem(blockName, modifiers, baseClass)}>{props.children}</span>;
}
