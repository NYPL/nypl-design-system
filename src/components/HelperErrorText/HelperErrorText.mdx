import { ArgTypes, Canvas, Description, Meta } from "@storybook/blocks";

import * as HelperErrorTextStories from "./HelperErrorText.stories";

<Meta of={HelperErrorTextStories} />

# HelperErrorText

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.0.10`   |
| Latest            | `1.5.0`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [HTML and JSX Children](#html-and-jsx-children)
- [Invalid State](#invalid-state)

## Overview

<Description of={HelperErrorTextStories} />

## Component Props

<Canvas of={HelperErrorTextStories.WithControls} />

<ArgTypes of={HelperErrorTextStories.WithControls} />

## Accessibility

In the Reservoir Design System (DS), the `HelperErrorText` component is always
used as a child component when composing more complex components.

In the case of form components, the `HelperErrorText` component is associated
with a related input field by using the `aria-describedby` attribute in the
associated element. This pattern ensures that the content of the
`HelperErrorText` component is made available to screenreaders.

This component will always render at least an empty `<div>` element with an
`aria-live="polite"` attribute set even if no text content is passed. This is to
ensure that the accessibility guideline for dynamic content is met. One expected
use case of the `HelperErrorText` component is to dynamically render error text
when a form field is invalid. In that scenario, the dynamically updated error
text will be read by assistive technologies.

The example below is shows a `TextInput` component. The `TextInput` component
uses the `HelperErrorText` component to render the red-colored `"This is error text :("` text below the input field and that text element is referenced with
the `aria-describedby` attribute on the HTML input element. This type of
association is handled automatically by all DS `Form Elements` components.

<Canvas of={HelperErrorTextStories.TextInputExample} />

### ariaAtomic

The `ariaAtomic` prop sets the `aria-atomic` HTML attribute. This allows the
entire DOM element to be presented to assistive technologies. When it is set to
`false`, only additions or removals will be read by assistive technologies.

By default, this prop is set to `true`.

### ariaLive

The `ariaLive` prop sets the `aria-live` HTML attribute. This determines the
priority of when text updates in this component should be read to users by
assistive technologies. The possible values are the same as the native HTML
`aria-live` attribute: `"off"`, `"polite"`, and `"assertive"`.

The `"off"` value indicates that updates should not be announced and when this
value is passed, the `aria-live` attribute will not be added at all. The
`"polite"` value indicates that updates will be announced at the next available
time slot. The `"assertive"` value indicates that updates should be announced
immediately, but this is not recommended as it can interruptive to the user.

In order to meet the accessibility guidelines for dynamic content updates, this
attribute is set in the wrapper `<div>` element that is always rendered by the
`HelperErrorText` component.

By default, this prop is set to `"polite"`.

<Canvas of={HelperErrorTextStories.AriaControls} />

Resources:

- [MDN aria-describedby Attribute](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby)
- [MDN ARIA live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)
- [W3C WAI-ARIA 1.1 - aria-live (property)](https://www.w3.org/TR/wai-aria-1.1/#aria-live)
- [W3C ARIA19: Using ARIA role=alert or Live Regions to Identify Errors](https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html)
- [DigitalA11y WAI-ARIA:aria-live (Property)](https://www.digitala11y.com/aria-live-properties/)

## HTML and JSX Children

The `HelperErrorText` component can render any HTML element or React component
through the `text` prop.

<Canvas of={HelperErrorTextStories.HTMLChildren} />

## Invalid State

Set the `isInvalid` prop to `true` to render the passed text content with the
NYPL "invalid" styling.

<Canvas of={HelperErrorTextStories.InvalidState} />
