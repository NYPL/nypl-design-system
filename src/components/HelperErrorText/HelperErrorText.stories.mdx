import {
  ArgsTable,
  Canvas,
  Description,
  Meta,
  Story,
} from "@storybook/addon-docs";
import { withDesign } from "storybook-addon-designs";

import HelperErrorText from "./HelperErrorText";
import Link from "../Link/Link";
import Text from "../Text/Text";
import { getCategory } from "../../utils/componentCategories";
import DSProvider from "../../theme/provider";

<Meta
  title={getCategory("HelperErrorText")}
  component={HelperErrorText}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url: "",
    },
    jest: ["HelperErrorText.test.tsx"],
  }}
  argTypes={{
    ariaAtomic: {
      control: false,
      table: { defaultValue: { summary: true } },
    },
    ariaLive: {
      table: { defaultValue: { summary: "polite" } },
    },
    children: { table: { disable: true } },
    className: { control: false },
    id: { control: false },
    isInvalid: {
      table: { defaultValue: { summary: false } },
    },
  }}
/>

# HelperErrorText

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.0.10`   |
| Latest            | `1.0.6`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [With HTML Children](#with-html-children)
- [Invalid State](#invalid-state)

## Overview

<Description of={HelperErrorText} />

## Component Props

<Canvas withToolbar>
  <Story
    name="HelperErrorText with Controls"
    args={{
      ariaAtomic: undefined,
      ariaLive: "polite",
      className: undefined,
      id: "helperErrorText-id",
      isInvalid: false,
      text: "This is the helper text!",
    }}
  >
    {(args) => (
      <HelperErrorText
        {...args}
        text={args.isInvalid ? "This is the error text :(" : args.text}
      />
    )}
  </Story>
</Canvas>

<ArgsTable story="HelperErrorText with Controls" />

## Accessibility

In the "invalid" state when `isInvalid` is set to true, the text passed to the
`invalidText` prop will be presented to screen readers be default. If you need
to turn this off, set the `ariaLive` prop to `"off"`. This is used in DS "form"
components for invalid form submissions or incorrect values entered.

### ariaAtomic

The `ariaAtomic` prop sets the `aria-atomic` attribute and it is set to `true`
by default. This allows the entire DOM element to be presented to assistive
technologies. When it is set to `false`, only additionals or removals will be
read by assistive technologies.

### ariaLive

The `ariaLive` prop sets the `aria-live` attribute; `"polite"` is set by default
in the "invalid" state and `"off"` otherwise. This sets the priority of when the
text in this component should be read to users by assistive technologies. The
`"off"` value indicates that the content should not be presented, `"polite"` that
it will be announced at the next available time slot, and `"assertive"` that it
should be announced immediately.

<Canvas>
  <Story
    name="With Aria Controls"
    args={{
      ariaAtomic: true,
      ariaLive: "polite",
      isInvalid: true,
    }}
  >
    {(args) => (
      <HelperErrorText
        {...args}
        text="Only invalid text can be read to screen readers with the appropriate aria-atomic and aria-live props."
      />
    )}
  </Story>
</Canvas>

Resources:

- [MDN ARIA live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)
- [W3C WAI-ARIA 1.1 - aria-live (property)](https://www.w3.org/TR/wai-aria-1.1/#aria-live)
- [W3C ARIA19: Using ARIA role=alert or Live Regions to Identify Errors](https://www.w3.org/TR/WCAG20-TECHS/ARIA19.html)
- [DigitalA11y WAI-ARIA:aria-live (Property)](https://www.digitala11y.com/aria-live-properties/)

## With HTML Children

The `HelperErrorText` component can render any React component or HTML element
through the `text` prop.

<Canvas>
  <Story name="With HTML Children">
    <HelperErrorText
      text={
        <>
          This first example uses an HTML anchor element for{" "}
          <a href="#">a link</a>.
        </>
      }
    />
    <HelperErrorText
      text={
        <Text>
          This second example uses Reservoir Design System (DS) components, such
          as the <Link href="#">`Link`</Link> component, and the `Text`
          component.
        </Text>
      }
    />
  </Story>
</Canvas>

## Invalid State

Set the `isInvalid` prop to `true` to render the passed text in the
NYPL "invalid" styling.

<Canvas>
  <Story name="Invalid State">
    <HelperErrorText isInvalid text="This is the error text :(" />
  </Story>
</Canvas>
