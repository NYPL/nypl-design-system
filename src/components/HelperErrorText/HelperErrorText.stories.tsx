import * as React from "react";

import HelperErrorText from "./HelperErrorText";
import { boolean } from "@storybook/addon-knobs";

export default {
  title: "HelperErrorText",
  component: HelperErrorText,
};

export const helperTextWithLink = () => (
  <HelperErrorText id="helperTextWithLink" isError={false}>
    {"If you're unsure of your size, please view the "}
    <a href="#sizing-chart">Sizing Chart</a>
  </HelperErrorText>
);

export const errorText = () => (
  <HelperErrorText id="helpertextWithError" isError={true}>
    Error Message
  </HelperErrorText>
);

export const longHelperText = () => (
  <HelperErrorText id="longHelperText" isError={false}>
    {"Souffl√© jujubes cupcake chocolate bar topping. Lollipop sesame snaps " +
      "tootsie roll liquorice sweet chupa chups. Gummi bears tart sweet roll " +
      "tiramisu biscuit cookie cake. Cookie chocolate jelly-o topping. Halvah " +
      "bear claw wafer cupcake tiramisu ice cream tart gummi bears. Lemon drops " +
      "chocolate cake croissant lemon drops gummies ice cream sugar plum."}
  </HelperErrorText>
);

export const withAriaProps = () => (
  <HelperErrorText
    id="withAriaProps"
    isError={true}
    ariaAtomic={boolean("ariaAtomic", false)}
    ariaLive="assertive"
  >
    {"We'll be making a water landing, "}
    <span>{"but that's OK because this is a seaplane."}</span>
  </HelperErrorText>
);
