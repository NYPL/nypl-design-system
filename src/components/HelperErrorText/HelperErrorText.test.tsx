import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";

import HelperErrorText from "./HelperErrorText";

describe("HelperErrorText Test", () => {
  it("Renders HelperErrorText", () => {
    const container = Enzyme.mount(
      <HelperErrorText id="helperTextWithLink" isError={false}>
        Text
      </HelperErrorText>
    );
    expect(container.exists("#helperTextWithLink")).to.equal(true);
  });

  it("Has 'error' modifier if error is passed", () => {
    const container = Enzyme.mount(
      <HelperErrorText id="helperTextWithLink" isError={true}>
        Text
      </HelperErrorText>
    );
    expect(container.exists("#helperTextWithLink")).to.equal(true);
  });

  it("Has aria-live and aria-atomic properties when errored", () => {
    const container = Enzyme.mount(
      <HelperErrorText id="helperTextWithLink" isError={true}>
        Text
      </HelperErrorText>
    );
    expect(container.find("div").props()["aria-live"]).to.equal("polite");
    expect(container.find("div").props()["aria-atomic"]).to.equal(true);
  });

  it("Accepts an aria-atomic value of false", () => {
    let container = Enzyme.mount(
      <HelperErrorText id="helperTextWithLink" isError={true}>
        Text
      </HelperErrorText>
    );
    // The default is "true".
    expect(container.find("div").props()["aria-atomic"]).to.equal(true);

    container = Enzyme.mount(
      <HelperErrorText
        id="helperTextWithLink"
        isError={true}
        ariaAtomic={false}
      >
        <p>
          This is static <span>but this part changes often!</span>
        </p>
      </HelperErrorText>
    );
    // But the prop accepts false in case only part of the helper text
    // should only be read instead of the whole region.
    expect(container.find("div").props()["aria-atomic"]).to.equal(false);
  });
});
