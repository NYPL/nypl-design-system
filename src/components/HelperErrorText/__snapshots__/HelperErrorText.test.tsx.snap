// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelperErrorText Renders the UI snapshot correctly 1`] = `
<div
  aria-atomic={true}
  aria-live="off"
  className=" css-0"
  data-isinvalid={false}
  id="basic"
>
  Text
</div>
`;

exports[`HelperErrorText Renders the UI snapshot correctly 2`] = `
<div
  aria-atomic={true}
  aria-live="polite"
  className=" css-0"
  data-isinvalid={true}
  id="invalid"
>
  Text
</div>
`;
